keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1428 @@
+node).  Storing the private keys in an HSM closes this loophole, but will require
+an additional interface for kubelet to expose remote signing capabilities to workloads.
+
+**Be 100% compatible with software in the field today**: The primary example in
+this KEP is that the private key and certificate chain are delivered in a single
+credential-bundle file.  This is based on implementation and operation
+experience of a CSI-based certificates solution that delivered private keys and
+certificate chains in separate files, as expected by most commodity webservers
+today.  While this is an attractive idea, automatically rotating these separate
+files is fraught.  Even if both files' content is updated atomically, skewed
+reads in the workload can result in hard-to-diagnose errors if the workload
+reads the key before rotation and the certificate chain after rotation (or
+vice-versa).
+
+**Support private keys shared across multiple pods**: Sharing private keys and
+certificates across multiple pods may be required in certain applications.
+However, this use case is well-served by Secret projected volume sources.
+
+**Support human-in-the-loop approval for certificate requests**: This KEP
+assumes that PodCertificateRequest issuance is fully-automated, because this
+process blocks pod startup for pods that use PodCertificate projected volume
+sources.  PodCertificateRequests do not separate the approval and issuance
+steps, unlike CertificateSigningRequests.
+
+**Standardize on using the SPIFFE SVID format in core Kubernetes**: The
+certificates issued by `kubernetes.io/kube-apiserver-client-pod` can be made
+compatible with SPIFFE in the future, since they currently don't include any
+Subject Alternate Name entries.
+
+**Support identity delegation or impersonation**: We do not intend to support
+issuing certificates that workloads can use as intermediate CA certificates in
+order to issue further certificates, nor do we intend to support one privileged
+workload minting certificates on behalf of other workloads.
+
+## Design Details
+
+### PodCertificateRequest Resource
+
+PodCertificateRequest is a new API type in certificates.k8s.io/v1alpha1.  It's a
+stripped-down version of CertificateSigningRequest that is tailored directly to
+the pod certificate issuance use case.
+
+PodCertificateRequest is fundamentally a statement that "Pod X is requesting a
+certificate from Signer Y".  The signer is in complete control of the form of
+the issued certificate.  As such, PodCertificateRequest objects don't contain a
+typical X.509 CSR.  Instead, they contain:
+* A public key, PKIX-serialized.  This key must be an RSA key with modulus size
+  3072 or 4096, an ECDSA key with curve P256 or P384, or an ED25519 key.
+* A proof of possession of the corresponding private key.
+  * If the key is an RSA key, then the signature is over the ASCII bytes of the
+	  pod UID, using RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as implemented
```

Comment:
I don't know how important this is --- the primary security guarantees here come from ensuring that only the correct kubelet can request a certificate for a given pod.  The proof of possession is just to prevent (implausible) accidents, as far as I can tell from https://www.rfc-editor.org/rfc/rfc4211#appendix-C

Reply:
And https://www.rfc-editor.org/rfc/rfc4211#section-4 only says that the proof of possession should sign over some representation of the identity being requested.  It doesn't include a timestamp.

Reply:
Okay per my reading of that RFC, I think the pod UID is sufficient when combined with all the other node admission checks (i.e. we know that this exact pod is scheduled on this exact node with this exact SA in the same namespace).

If a different node tried to replay the PoP, it would fail because that pod isn't scheduled to it.